File: android/app/src/main/java/com/dynops/bcwms/feature/QcErrorParser.kt
package com.dynops.bcwms.feature

/**
 * Quality-related API error parser.
 *
 * MS Quality Management + DOPSWHS workflow events block Picks / Put-Aways
 * when a referenced lot/serial/package has an open Quality Inspection that
 * has not yet finished with a passing result. BC surfaces these blocks as
 * error responses on the bound action call.
 *
 * This helper detects the QC-block signature in the error text and pulls
 * out the inspection no. when present, so the mobile UI can show a
 * friendly banner ("Lot blocked by QC inspection INS-001 — bekle veya
 * Quality modülünden tamamla") instead of a raw HTTP error.
 *
 * Used by:
 * - PickingModule.kt (assignToMe / register / markShort actions)
 * - PutAwayShipModules.kt (Put-Away register, Shipment post)
 */
object QcErrorParser {

    private val INSPECTION_NO_REGEX = Regex("""\b(?:INS|QI|Q|QM)[\-_]?\d{1,8}\b""")
    private val QC_KEYWORDS = listOf("quality", "inspection", "qc", "blocked", "blok")

    /** True if the error text matches a Quality / Inspection block. */
    fun isQcBlocked(errMsg: String?): Boolean {
        if (errMsg.isNullOrBlank()) return false
        val lower = errMsg.lowercase()
        val hasKeyword = QC_KEYWORDS.count { lower.contains(it) } >= 2
        return hasKeyword || INSPECTION_NO_REGEX.containsMatchIn(errMsg)
    }

    /** Inspection no. if mentioned, else null. */
    fun extractInspectionNo(errMsg: String?): String? {
        if (errMsg.isNullOrBlank()) return null
        return INSPECTION_NO_REGEX.find(errMsg)?.value
    }

    /**
     * UI-friendly status line for a QC-blocked error. Falls back to the
     * original error if no inspection no. is found.
     */
    fun friendlyStatus(errMsg: String?, httpCode: Int = 0): String {
        val raw = errMsg ?: "Bilinmeyen hata"
        if (!isQcBlocked(raw)) return "HATA: $raw (HTTP $httpCode)"
        val ins = extractInspectionNo(raw)
        val tail = if (ins != null) "Quality Inspection $ins tamamlanmalı." else "Açık bir Quality Inspection bekliyor."
        return "🔬 QC BLOCK: Lot/Serial bloklu — $tail (HTTP $httpCode)"
    }
}

File: web/src/lib/qcErrorParser.ts
/**
 * Quality-related API error parser (web counterpart of QcErrorParser.kt).
 *
 * MS Quality Management + DOPSWHS workflow events block Picks / Put-Aways
 * when a referenced lot/serial/package has an open Quality Inspection that
 * has not yet finished with a passing result. BC surfaces these blocks as
 * error responses on the bound action call.
 *
 * The web UI uses this helper to detect QC-block errors and render a
 * friendly banner with a deep-link to the Quality module instead of a raw
 * HTTP error string.
 */

const INSPECTION_NO_REGEX = /\b(?:INS|QI|Q|QM)[-_]?\d{1,8}\b/;
const QC_KEYWORDS = ["quality", "inspection", "qc", "blocked", "blok"];

export function isQcBlocked(errMsg: string | null | undefined): boolean {
  if (!errMsg) return false;
  const lower = errMsg.toLowerCase();
  const hits = QC_KEYWORDS.filter((k) => lower.includes(k)).length;
  return hits >= 2 || INSPECTION_NO_REGEX.test(errMsg);
}

export function extractInspectionNo(errMsg: string | null | undefined): string | null {
  if (!errMsg) return null;
  const m = errMsg.match(INSPECTION_NO_REGEX);
  return m ? m[0] : null;
}

export function friendlyQcStatus(errMsg: string | null | undefined, httpCode = 0): string {
  const raw = errMsg ?? "Bilinmeyen hata";
  if (!isQcBlocked(raw)) return `HATA: ${raw} (HTTP ${httpCode})`;
  const ins = extractInspectionNo(raw);
  const tail = ins
    ? `Quality Inspection ${ins} tamamlanmalı.`
    : "Açık bir Quality Inspection bekliyor.";
  return `🔬 QC BLOCK: Lot/Serial bloklu — ${tail} (HTTP ${httpCode})`;
}
